### ISSUE TITLE
Plasma 6: Application Menu Does not Refresh List when Applications added/removed

### ISSUE THREAD

**Issue by Lyndeno** [NixOS/nixpkgs#292632]
### Describe the bug
When applications are added or removed, the menu stays with the current list from when I originally installed Plasma. Reboots or restarts does not refresh it.

### Steps To Reproduce
1. Install Plasma
2. Add or remove applications (in my case within nixos config)
3. Menu keeps old list

### Expected behavior
Application menu refreshed with current available applications

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by xNaxdy** [comment]
The problem seems to be that when a new `ksycoca6_.....` file is created in the `~/.cache`, plasma just completely ignores it and happily continues to use whichever one it has been using. A workaround (for now) is to manually run `rm ~/.cache/ksycoca*` (or add it to your activation scripts, lol) after every `nixos-rebuild`, and this should take care of things.

Still, it'd be nice to know if this is a NixOS issue, or an upstream issue. Feels like it could be either?

**Comment by reasonings** [comment]
How would I go about getting the kbuildsyscoca6 binary from inside the "activationScript"? I can't seem to find the correct ${whatever}

**Comment by reasonings** [comment]
> A workaround (for now) is to manually run `rm ~/.cache/ksycoca*` (or add it to your activation scripts, lol) after every `nixos-rebuild`, and this should take care of things.

When I add it to my activation scripts, it does remove it, but it never comes back. All icons go blank and the only way to get them back is to reboot.

**Comment by pbek** [comment]
Does a `plasmashell replace` help?

**Comment by K900** [comment]
So there are actually multiple issues here. Most KDE applications pull their knowledge of currently available apps from a shared cache called sycoca (for SYstem COnfiguration CAche), which is maintained by KService. Unfortunately, it does not work correctly on NixOS for multiple reasons:

1. The cache can be rebuilt in-process, by each application, but the name and contents of the cache depend on XDG_DATA_DIRS, which is different for many processes due to wrappers, so we end up with multiple unrelated copies of the cache, each maintained by a different application, with a different view of things. This isn't _necessarily_ an issue when the cache is updated correctly, except wasting a whole bunch of cycles.
2. The cache is updated by checking the mtimes on the target files, which for us are always 0, so the cache is effectively never refreshed unless nuked entirely. The solution to that is probably to make KService actually walk up the path until it finds a symlink to the store, and then check the mtime of _that_, but that's an extremely invasive change.

https://github.com/NixOS/nixpkgs/pull/292748 should unblock us for now by restoring the nuke-all-sycoca activation script from Plasma 5, but we'll definitely want to look into some kind of a better solution going forward, and probably get upstream involved.

**Comment by K900** [comment]
Can people affected by this please test https://github.com/NixOS/nixpkgs/pull/292748 ?

**Comment by JonathanBrouwer** [comment]
For me running `xdg-desktop-menu forceupdate` fixes the problem as well, so as a workaround I added that to my rebuild script

**Comment by ShalokShalom** [comment]
I had the same issue with NeoChat and was informed, that [this fix](https://github.com/NixOS/nixpkgs/pull/292483) corrects the issue and indeed, after an update and reboot, it showed up.

**Comment by NovaViper** [comment]
> Can people affected by this please test #292748 ?

I'm still plagued by this issue, neither installing kcmutils nor the nuke script functions on my system (My lock file has nixpkgs at 73d699a6ff1e83df3fd6c1e60931e13667b8ae14). My flake: https://github.com/NovaViper/NixConfig

**Comment by sdellysse** [comment]
I'm still seeing this issue. I run `nixos-rebuild switch` and now when I click on apps in my KDE "Start Menu", nothing happens. Running latest nixos-stable, updated and rebuilt this morning

**Comment by pbek** [comment]
Do you have this issue in the "Favorites" part of the menu? (because I even have that issue under Plasma 5)

**Comment by Shawn8901** [comment]
I can confirm for the Favorites. Normal menu items do work for me in the meantime totally fine.

**Comment by pbek** [comment]
Ok, I guess then that's a general Plasma5/6 issue on NixOS. 😅
Is this worth a new issue, @K900?

**Comment by K900** [comment]
I have not observed this issue on 6, at least.

**Comment by pbek** [comment]
Strange... I just tested it after a QOwnNotes update in the morning on a plasma 6 Laptop. 🤔
I can search for it or use the normal menu entry from the different categories, but the favorite entry can't be clicked.

**Comment by micmalti** [comment]
I've noticed the same issue in KRunner when doing an app search. Currently on Plasma 6.0.5.

**Comment by tazihad** [comment]
Fedora 41 KDE plasma has this issue. Although restoring it brings the icons. So, I guess nix need to refresh KDE plasma start menu.  
![Screenshot_20240930_181306](https://github.com/user-attachments/assets/41362601-db6a-42df-afbe-68d02981dc89)

**Comment by Red-M** [comment]
I've got this issue and I installed my NixOS within the last week.
Using stable and I get the same issues, the favourites section of plasma doesn't launch anything and I also can't edit them after a `nixos-rebuild` if the packages get changed.

**Comment by K900** [comment]
This should be mostly worked around in 24.11.

**Comment by yodaldevoid** [comment]
I am still seeing this occur on unstable at the moment. The original "fix" of running `rm ~/.cache/ksycoca*` when I need to force an update has worked for me. `xdg-desktop-menu forceupdate` seems to do nothing. I am unsure of what to do to debug it past that.

**Comment by pbek** [comment]
For me, this issue still only applies to the `Favorites` in the Application Menu on NixOS Unstable. But neither running `rm ~/.cache/ksycoca*` nor `xdg-desktop-menu forceupdate` seem to help for me. 🤔

**Comment by SourLemonJuice** [comment]
How about to run `kbuildsycoca6` or `kbuildsycoca5` as user? Google's AI and this forum tell me it can force rebuild KDE's menu cache, and it works on my Archlinux.\
[How to reload application menu in Plasma 5](https://forum.kde.org/viewtopic.php%3Ff=289&t=129411.html)

In my case, I need relogin or even reboot to see the new installed app when using nix package manager. Rebuild the cache can fix it.

**Comment by pbek** [comment]
I tried `kbuildsycoca6` under plasma6 after the favorites were broken, and it didn't fix them. I tried `kbuildsycoca6 --menutest` and it didn't show that it would touch favorites. 😬

**Comment by RoGreat** [comment]
> For me, this issue still only applies to the `Favorites` in the Application Menu on NixOS Unstable. But neither running `rm ~/.cache/ksycoca*` nor `xdg-desktop-menu forceupdate` seems to help for me. 🤔

I don't know if you are having the same issue as me (what Red-M said last year). Have to restart plasmashell `sudo systemctl restart --machine=<user>@.host --user plasma-plasmashell.service` to fix the issue after a `nixos-rebuild switch`. Put in a bug report here: https://bugs.kde.org/show_bug.cgi?id=508368.

Even more surprising that this bug was reported nearly a decade ago and is still happening now: https://github.com/NixOS/nixpkgs/issues/10579.

**Comment by LunNova** [comment]
Does the following cause it to update without needing to restart plasma-shell?

```
$ nix shell nixpkgs#dbus nixpkgs#kdePackages.qttools
qdbus org.kde.KWin /KWin reconfigure || echo "KWin reconfigure failed"
for i in {0..10}; do
  dbus-send --type=signal /KGlobalSettings org.kde.KGlobalSettings.notifyChange int32:$i int32:0 || echo "KGlobalSettings.notifyChange failed"
done
```
